vllm/model_executor/layers/mamba/mamba_mixer2.py

```diff
@@ -493,10 +494,17 @@ def forward_cuda(
                 dt_limit=(0.0, float("inf")),
             )
 
-            # update ssm states
-            # - varlen state is a (batch, nheads, headdim, dstate) tensor
-            for i, idx in enumerate(mamba_cache_params.state_indices_tensor):
-                mamba_cache_params.ssm_state[idx].copy_(varlen_state[i])
+            # vectorized ssm state update using vmap
+            # the 1d state_indices_tensor needs to be unsqueezed to avoid vmap
+            # limitation which doesn't allow use of `item()`
+            # Note: the lambda capture can happen where ssm_state is initialized
+            #       instead of here
+            batched_copy = torch.vmap(
+                lambda idx, source_state: mamba_cache_params.ssm_state[
+                    idx].copy_(source_state))
```

Comment:
This might be handy to have as a method of `MambaCacheParams` in `mamba_cache.py`

Reply:
@tlrmchlsmth could you clarify if you mean to have this logic as a member function of `MambaCacheParams` ?

Reply:
That's right, that's what I meant, although I don's see a way to factor out commonality between `batched_copy` and `batched_zero_init_func` so I'm not sure it would clean anything up.

>            # Note: the lambda capture can happen where ssm_state is initialized
>            #       instead of here
Is there some overhead that we should try to avoid here?

Reply:
The current lambda capture code is safe. The comment is just theorizing about removing redundancy. I don't know this part well enough yet. Attempting to "optimize" may introduce bugs. I'd leave it as is for now.